pkg/probe/grpc/grpc_test.go

```diff
@@ -0,0 +1,146 @@
+	grpchealth "google.golang.org/grpc/health/grpc_health_v1"
+
+	"k8s.io/kubernetes/pkg/probe"
+)
+
+func TestNew(t *testing.T) {
+	t.Run("Should: implement Probe interface", func(t *testing.T) {
+		s := New()
+		assert.Implements(t, (*Prober)(nil), s)
+	})
+}
+
+type successServerMock struct {
+}
+
+func (s successServerMock) Check(context.Context, *grpchealth.HealthCheckRequest) (*grpchealth.HealthCheckResponse, error) {
+	return &grpchealth.HealthCheckResponse{
+		Status: grpchealth.HealthCheckResponse_SERVING,
+	}, nil
+}
+
+func (s successServerMock) Watch(_ *grpchealth.HealthCheckRequest, stream grpchealth.Health_WatchServer) error {
+	return stream.Send(&grpchealth.HealthCheckResponse{
+		Status: grpchealth.HealthCheckResponse_SERVING,
+	})
+}
+
+type errorTimeoutServerMock struct {
+}
+
+func (e errorTimeoutServerMock) Check(context.Context, *grpchealth.HealthCheckRequest) (*grpchealth.HealthCheckResponse, error) {
+	time.Sleep(time.Second * 4)
+	return &grpchealth.HealthCheckResponse{
+		Status: grpchealth.HealthCheckResponse_SERVING,
+	}, nil
+}
+
+func (e errorTimeoutServerMock) Watch(_ *grpchealth.HealthCheckRequest, stream grpchealth.Health_WatchServer) error {
+	time.Sleep(time.Second * 4)
+	return stream.Send(&grpchealth.HealthCheckResponse{
+		Status: grpchealth.HealthCheckResponse_SERVING,
+	})
+}
+
+type errorNotServeServerMock struct {
+}
+
+func (e errorNotServeServerMock) Check(context.Context, *grpchealth.HealthCheckRequest) (*grpchealth.HealthCheckResponse, error) {
+	return &grpchealth.HealthCheckResponse{
+		Status: grpchealth.HealthCheckResponse_NOT_SERVING,
+	}, nil
+}
+
+func (e errorNotServeServerMock) Watch(_ *grpchealth.HealthCheckRequest, stream grpchealth.Health_WatchServer) error {
+	return stream.Send(&grpchealth.HealthCheckResponse{
+		Status: grpchealth.HealthCheckResponse_NOT_SERVING,
+	})
+}
+
+func TestGrpcProber_Probe(t *testing.T) {
+	t.Run("Should: failed but return nil error because cant find host", func(t *testing.T) {
+		s := New()
+		p, _, err := s.Probe("", "", 32, time.Second, grpc.WithInsecure(), grpc.WithBlock())
+		assert.Equal(t, probe.Failure, p)
+		assert.Equal(t, nil, err)
+	})
+	t.Run("Should: return nil error because server response not served", func(t *testing.T) {
+		s := New()
+		lis, _ := net.Listen("tcp", ":10413")
+		port := lis.Addr().(*net.TCPAddr).Port
+		grpcServer := grpc.NewServer()
+		defer grpcServer.Stop()
+		grpchealth.RegisterHealthServer(grpcServer, &errorNotServeServerMock{})
+		go func() {
+			_ = grpcServer.Serve(lis)
+		}()
+		// take some time to wait server boot
+		time.Sleep(2 * time.Second)
+		p, o, err := s.Probe("0.0.0.0", "", port, time.Second, grpc.WithInsecure())
+		assert.Equal(t, probe.Failure, p)
+		assert.Equal(t, nil, err)
+		assert.Equal(t, "GRPC probe failed with status: NOT_SERVING", o)
+	})
+	t.Run("Should: return nil-error because server not response in time", func(t *testing.T) {
+		s := New()
+		lis, _ := net.Listen("tcp", ":10414")
```

Comment:
port 0?

Reply:
Grpc server will use random port, if we set `lis, _ := net.Listen("tcp", ":0")`

Reply:
that is what we want for the test, otherwise this can collide , 
you get the port for the test in the next line `port := lis.Addr().(*net.TCPAddr).Port`
am I missing something?

Reply:
I see, add a new test case for `lis, _ := net.Listen("tcp", ":0")`.

Reply:
nope, but maybe I'm missing something, why do you need to hardcode the port with the risk of failing the test becaues of collision,  if the OS can give you a free random port.
The test doesn't need to know the hardcoded port, does it? it can use the one we get

Reply:
em, `port` only refer in `p, _, err := s.Probe("0.0.0.0", "", port, time.Second*2, grpc.WithInsecure())` , i think maybe you are right